Add InterBlock loop/epilogue analysis

andcarminati opened the pull request:
Implement loop/epilogue analysis to reduce pessimism related to the propagation of loop latencies and resource conflicts to epilogue.
This PR adds:
* MemoryEdges update to ignore bundles.
* Latency analysis using a DDG.
* Conflict analysis using two scoreboards.
* Code refactoring.
* Test additions/updates.

andcarminati committed:
[AIE2] Change MemoryEdges to ignore bundles

This change is necessary to run correctly mutations after bundling.

andcarminati committed:
[AIE2] Add InterBlock loop/epilogue analysis

Implement loop/epilog analysis to reduce pessimism related to the
propagation of loop latencies and resource conflicts to the epilogue.

This commit adds:
  * Latency analysis using a DDG.
  * Conflict analysis using two scoreboards.
  * Code refactoring.
  * Test additions/updates.

andcarminati commented:
Benchmark evaluation:

With option `--aie-loop-min-tripcount=4`

First results are without the change and the second are with.
```
+----------------+-------------------+------------------+--+-------------------+------------------+--+--------------------+-------------------+
|     Design     | Total Cycle Count | Total Insn Count |  | Total Cycle Count | Total Insn Count |  | % Dec. Cycle Count | % Dec. Insn Count |
+----------------+-------------------+------------------+--+-------------------+------------------+--+--------------------+-------------------+
| GEMM_bf16_0    |             10994 |             4670 |  |             10898 |             4574 |  | 0,87%              | 2,06%             |
| GEMV_0         |              3163 |              689 |  |              3163 |              689 |  | 0,00%              | 0,00%             |
| Mul2D_0        |              4807 |             1643 |  |              4807 |             1643 |  | 0,00%              | 0,00%             |
| AvgPool2D_0    |              5153 |             3299 |  |              5153 |             3299 |  | 0,00%              | 0,00%             |
| Pad2D_0        |              4522 |             2190 |  |              4522 |             2190 |  | 0,00%              | 0,00%             |
| MaxPool2D_0    |              3832 |             2483 |  |              3832 |             2483 |  | 0,00%              | 0,00%             |
| GEMV_1         |              1969 |              587 |  |              1969 |              587 |  | 0,00%              | 0,00%             |
| Add2D_0        |              8157 |             3478 |  |              8157 |             3478 |  | 0,00%              | 0,00%             |
| Conv2D_0       |             20649 |            10943 |  |             20239 |            10533 |  | 1,99%              | 3,75%             |
| Conv2D_ReLU_0  |              3391 |             1716 |  |              3331 |             1656 |  | 1,77%              | 3,50%             |
| Conv2D_LReLU_0 |              4768 |             3094 |  |              4708 |             3034 |  | 1,26%              | 1,94%             |
+----------------+-------------------+------------------+--+-------------------+------------------+--+--------------------+-------------------+


```

martien-de-jong commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,168 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+  // We can only analyze non-empty epilogue blocks because we need
+  // to build a DDG, which is not possible.
+  // For empty ones, we need to be conservative because we are not aware of
+  // content of epilogues' sucessor.
```
nit: successor

martien-de-jong replied:
I'm wondering about the above comment, because there's no real distinction between 'empty' or 'too short'. 
The DAG is created anyway, and the RegionEndEdges should push ExitSU out by the right amount. I guess the problem is because we will don't inspect edges to ExitSU

andcarminati replied:
This is because we are handling vectors that cannot be empty. This prevents vectors' assertion's from kick in.

martien-de-jong commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,168 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+        } else {
+          // We need to skip instructions like debug ones.
+          auto DepthIt = Depths.find(Succ->getInstr());
+          if (DepthIt == Depths.end()) {
```
Curious: since we only enter instructions from bundles, I don't expect any debug instructions. Do we ever take this path?

andcarminati replied:
You are right, it is safe to not expect debug instructions here.

martien-de-jong commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,168 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+  // instruction and start the resource verification from this point, tracking
+  // the number of NOPS.
+  int NopCounter = ExistingLatency;
+  for (NopCounter = 0; NopCounter < ExistingLatency; ++NopCounter)
```
nit: I would have `int NopCounter=0;` and leave the first for-clause empty, or write it as a while loop. That while loop will look nicely similar to the one below.

martien-de-jong reviewed (dismissed):
Nothing wrong.  Just a few nits you may want to consider.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -32,8 +32,56 @@ static cl::opt<bool>
+
+void emitBundlesInScoreboardDelta(
+    const std::vector<MachineBundle> &Bundles,
+    ResourceScoreboard<FuncUnitWrapper> &Scoreboard, int &Cycle,
```
Nit: Using the name `Cycle` is confusing me a bit, as this does not correspond to an actual emission cycle of the instructions we are emitting. This is really just a Delta from the `CurrCycle` of `Scoreboard`, and when that Delta reaches `0`, we know there's no need to replay more instructions and identify resource conflicts.

Maybe just rename to `Delta`, and add a comment where the `break` is to explain why we can exit?

andcarminati replied:
I updated this point, thank you!

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,162 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+  // Check cross-boundary latencies.
+  int Height = 1;
+  for (auto &Bundle : reverse(LoopBS.getBottom().Bundles)) {
+    for (auto *MI : Bundle.getInstrs()) {
```
Nit: Rename to `LoopMI` or `PreBoundaryMI`

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,162 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+  auto AddRegionToEdges = [&](const Region &R) {
+    for (auto &Bundle : R.Bundles) {
+      for (MachineInstr *MI : Bundle.getInstrs()) {
+        Depths[MI] = Depth;
```
Nit: The term `Depth` is extremely overloaded and it's hard for me to keep track of all the meanings. Here I think we use it to access the distance from the "pre-boundary" region. Maybe name it `DistFromLoopEntry` or `DistFromFirstBundle`?

andcarminati replied:
I updated the named to reflect better the semantics.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,162 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+          // We can consider the maximum depth+1 as depth of the
+          // ExitSU.
+          Distance = Depth - Depths[MI];
+          Latency = AIE::maxLatency(MI, *TII, *ItinData, false);
```
Do we still need the special casing now that `Edges.buildEdges();` also runs the `RegionEndEdges` mutator? I would expect that the latency is already correctly computed.

andcarminati replied:
In this case, it is necessary the identification of the boundary node because there is no instruction associated (no mapping). But, as the mutation is running now, I simplified the code to use the already computed latency.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +407,162 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+  AddRegionToEdges(LoopBS.getBottom());
+  Edges.markBoundary();
+  // Second part is the epilogue itself
+  AddRegionToEdges(EpilogueBS.getTop());
```
Nit: Maybe also add the `ExitSU` node of `EpilogueBS` to the map of distances so we can avoid the `if (Succ->isBoundaryNode())` corner case in the loop below.

andcarminati replied:
In the current implementation, this mapping maps instructions to depths, in this way we need to change the mapping logic as well.

gbossu replied:
What I meant is something like: `DistancesFromLoopEntry[EpilogueBS.getTop().getExitSU()] = DistFromLoopEntry;`, this way there is no `if (Succ->isBoundaryNode())` special casing in the loop below.

andcarminati replied:
Hi, I simplified a bit more the code.

gbossu reviewed (commented):
This looks good, I think the last remaining thing would be to fix the assertion that now triggers in the `RegionEndEdges` mutator for some benchmarks, and add a small reproducer test for it.

On top of that, please look at the Nit comments and address up to taste :) Generally, I'd be happy if we can get rid of the `if (Succ->isBoundaryNode())` special casing.

andcarminati committed:
[AIE2] Fix RegionEndEdges for bundled Instructions

Now, this mutation can be applied to already bundled instructions
without misclassification. It means that it is safe to run this mutation
with Epilogue scheduling, for example.

andcarminati committed:
[AIE2] Remove unnecessary cast to AIEScheduleDAGMI

If cast is necessary, MaxLatencyFinder will do it anyway.

andcarminati commented:
Hi @martien-de-jong and @gbossu, I addressed all the comments and suggestions. Benchmarks are running again with no problems. I you can take a look, it would be nice.

andcarminati commented:
Here the most recent results:

```

+----------------+---+------------+---------------+-------------+---+---+------------+---------------+-------------+---+---+-------------------+----------------+--------------------+
|       #        | # | Without EA |       #       |      #      | # | # |  With EA   |       #       |      #      | # | # |       Comp.       |       #        |         #          |
+----------------+---+------------+---------------+-------------+---+---+------------+---------------+-------------+---+---+-------------------+----------------+--------------------+
| Design         |   | Tot. PM S. | Tot. Cycle C. | Tot. Ins C. |   |   | Tot. PM S. | Tot. Cycle C. | Tot. Ins C. |   |   | % Diff Tot. PM S. | % Diff Tot. C. | % Diff Tot. Ins C. |
| GEMM_bf16_0    |   | 4160       | 10461         | 4160        |   |   | 4144       | 10365         | 4064        |   |   | -0,38%            | -0,92%         | -2,31%             |
| GEMV_0         |   | 3488       | 3126          | 650         |   |   | 3488       | 3126          | 650         |   |   | 0,00%             | 0,00%          | 0,00%              |
| Mul2D_0        |   | 2672       | 4857          | 1686        |   |   | 2672       | 4857          | 1686        |   |   | 0,00%             | 0,00%          | 0,00%              |
| AvgPool2D_0    |   | 3552       | 4159          | 2811        |   |   | 3552       | 4159          | 2811        |   |   | 0,00%             | 0,00%          | 0,00%              |
| Pad2D_0        |   | 3456       | 4512          | 2184        |   |   | 3456       | 4512          | 2184        |   |   | 0,00%             | 0,00%          | 0,00%              |
| MaxPool2D_0    |   | 3344       | 3798          | 2449        |   |   | 3344       | 3798          | 2449        |   |   | 0,00%             | 0,00%          | 0,00%              |
| GEMV_1         |   | 3488       | 1948          | 560         |   |   | 3488       | 1948          | 560         |   |   | 0,00%             | 0,00%          | 0,00%              |
| Add2D_0        |   | 3600       | 8157          | 3478        |   |   | 3600       | 8157          | 3478        |   |   | 0,00%             | 0,00%          | 0,00%              |
| Conv2D_0       |   | 5952       | 19018         | 10543       |   |   | 5920       | 18538         | 10063       |   |   | -0,54%            | -2,52%         | -4,55%             |
| Conv2D_ReLU_0  |   | 5328       | 3377          | 1704        |   |   | 5296       | 3317          | 1644        |   |   | -0,60%            | -1,78%         | -3,52%             |
| Conv2D_LReLU_0 |   | 6000       | 4767          | 3095        |   |   | 5968       | 4707          | 3035        |   |   | -0,53%            | -1,26%         | -1,94%             |
+----------------+---+------------+---------------+-------------+---+---+------------+---------------+-------------+---+---+-------------------+----------------+--------------------+


```

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +409,160 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+  while (Bottom.conflict(Top, Depth)) {
+    Bottom.advance();
+    NopCounter++;
+  }
```
Nit: That looks a lot like `AIEPostRASchedStrategy::handleRegionConflicts`, maybe there's code we can share.

andcarminati replied:
Hi @gbossu, I think it could be a bit complicated, because `handleRegionConflicts` uses two hazard recognizers, while we use just one, plus scoreboard comparison. I am afraid that a refactor could create a more confusing code.

gbossu replied:
The implementation is different, but ultimately the goal is the same, we want to insert NOPs until there is no resource hazard. We can do that in a follow-up PR.

gbossu reviewed (dismissed):
LGTM, I think it's great work!

Nit: I'd be even happier if we could remove the `if (Succ->isBoundaryNode())` special case in `getCyclesToRespectTiming()`, maybe by extending the map of depths, maybe by using a conditional assignment instead.

andcarminati commented:
> LGTM, I think it's great work!
> 
> Nit: I'd be even happier if we could remove the `if (Succ->isBoundaryNode())` special case in `getCyclesToRespectTiming()`, maybe by extending the map of depths, maybe by using a conditional assignment instead.

Thank you for this suggestion, I addressed this special case.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -369,14 +409,152 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+          continue;
+        const MachineInstr *PostBoundaryMI = Succ->getInstr();
+
+        int PostBondOrExitDist =
```
Nit `PostBoundOrExitDist`

gbossu reviewed (dismissed):
LGTM

martien-de-jong reviewed (approved):
LGTM
